File: 11_bermudan_swaption_tree_pricing/python/tree_pricer.py
"""
tree_pricer.py

Bermudan swaption pricing by backward induction.
"""

import numpy as np


class BermudanSwaptionTreePricer:
    """
    Bermudan swaption tree pricer.
    """

    def __init__(
        self,
        tree,
        bermudan_swaption
    ):

        self.tree = tree
        self.instrument = bermudan_swaption

        self.value_tree = None
        self.exercise_tree = None

    def is_exercise_time(
        self,
        t,
        tolerance=1e-8
    ):
        """
        Check whether t is an exercise date.
        """

        return np.any(
            np.abs(
                self.instrument.exercise_dates
                -
                t
            )
            <
            tolerance
        )

    def price(
        self
    ):
        """
        Backward induction.

        At each node:

            if exercise date:
                value = max(exercise value, continuation value)

            otherwise:
                value = continuation value
        """

        n_steps = self.tree.n_steps

        values = [
            np.zeros(step + 1)
            for step in range(n_steps + 1)
        ]

        exercise_flags = [
            np.zeros(step + 1, dtype=bool)
            for step in range(n_steps + 1)
        ]

        terminal_time = self.tree.times[-1]

        for node in range(n_steps + 1):

            r = self.tree.rates[-1][node]

            if self.is_exercise_time(
                terminal_time
            ):

                values[-1][node] = (
                    self.instrument.exercise_value(
                        model=self.tree.model,
                        exercise_time=terminal_time,
                        short_rate=r
                    )
                )

                exercise_flags[-1][node] = (
                    values[-1][node] > 0.0
                )

        for step in reversed(
            range(n_steps)
        ):

            t = self.tree.times[step]

            for node in range(step + 1):

                p_up = self.tree.probabilities[step][node]
                p_down = 1.0 - p_up

                df = self.tree.discount_factor_one_step(
                    step,
                    node
                )

                continuation_value = (
                    df
                    *
                    (
                        p_down
                        *
                        values[step + 1][node]
                        +
                        p_up
                        *
                        values[step + 1][node + 1]
                    )
                )

                node_value = continuation_value

                if self.is_exercise_time(t):

                    r = self.tree.rates[step][node]

                    exercise_value = (
                        self.instrument.exercise_value(
                            model=self.tree.model,
                            exercise_time=t,
                            short_rate=r
                        )
                    )

                    if exercise_value > continuation_value:

                        node_value = exercise_value
                        exercise_flags[step][node] = True

                values[step][node] = node_value

        self.value_tree = values
        self.exercise_tree = exercise_flags

        return values[0][0]

    def get_value_tree(
        self
    ):
        """
        Return value tree.
        """

        if self.value_tree is None:
            raise ValueError(
                "Run price() before accessing the value tree."
            )

        return self.value_tree

    def get_exercise_tree(
        self
    ):
        """
        Return exercise decision tree.
        """

        if self.exercise_tree is None:
            raise ValueError(
                "Run price() before accessing the exercise tree."
            )

        return self.exercise_tree

File: 11_bermudan_swaption_tree_pricing/cpp/TreePricer.cpp
#include "TreePricer.hpp"

#include <cmath>
#include <algorithm>

/*
    Constructor.
*/

BermudanSwaptionTreePricer::BermudanSwaptionTreePricer(
    HullWhiteBinomialTree* tree_,
    BermudanSwaption* instrument_
){

    tree = tree_;
    instrument = instrument_;
}

/*
    Check whether t is an exercise date.
*/

bool BermudanSwaptionTreePricer::isExerciseTime(
    double t,
    double tolerance
) const {

    for(double exerciseTime : instrument->exerciseDates){

        if(std::abs(exerciseTime - t) < tolerance){

            return true;
        }
    }

    return false;
}

/*
    Price by backward induction.

    At each node:

        continuation value
        exercise value
        option value = max(exercise, continuation)
        exercise flag
*/

double BermudanSwaptionTreePricer::price(){

    int nSteps =
        tree->nSteps;

    valueTree.clear();
    exerciseValueTree.clear();
    continuationValueTree.clear();
    exerciseFlagTree.clear();

    for(int step = 0; step <= nSteps; ++step){

        valueTree.push_back(
            std::vector<double>(
                step + 1,
                0.0
            )
        );

        exerciseValueTree.push_back(
            std::vector<double>(
                step + 1,
                0.0
            )
        );

        continuationValueTree.push_back(
            std::vector<double>(
                step + 1,
                0.0
            )
        );

        exerciseFlagTree.push_back(
            std::vector<bool>(
                step + 1,
                false
            )
        );
    }

    /*
        Terminal step.
    */

    double terminalTime =
        tree->times[nSteps];

    for(int node = 0; node <= nSteps; ++node){

        double r =
            tree->rates[nSteps][node];

        if(isExerciseTime(terminalTime)){

            double ex =
                instrument->exerciseValue(
                    *(tree->model),
                    terminalTime,
                    r
                );

            exerciseValueTree[nSteps][node] =
                ex;

            valueTree[nSteps][node] =
                ex;

            exerciseFlagTree[nSteps][node] =
                ex > 0.0;
        }
    }

    /*
        Backward induction.
    */

    for(int step = nSteps - 1; step >= 0; --step){

        double t =
            tree->times[step];

        for(int node = 0; node <= step; ++node){

            double pUp =
                tree->probabilities[step][node];

            double pDown =
                1.0 - pUp;

            double df =
                tree->discountFactorOneStep(
                    step,
                    node
                );

            double continuation =
                df
                *
                (
                    pDown
                    *
                    valueTree[step + 1][node]
                    +
                    pUp
                    *
                    valueTree[step + 1][node + 1]
                );

            continuationValueTree[step][node] =
                continuation;

            double nodeValue =
                continuation;

            if(isExerciseTime(t)){

                double r =
                    tree->rates[step][node];

                double exercise =
                    instrument->exerciseValue(
                        *(tree->model),
                        t,
                        r
                    );

                exerciseValueTree[step][node] =
                    exercise;

                if(exercise > continuation){

                    nodeValue =
                        exercise;

                    exerciseFlagTree[step][node] =
                        true;
                }
            }

            valueTree[step][node] =
                nodeValue;
        }
    }

    return
        valueTree[0][0];
}
